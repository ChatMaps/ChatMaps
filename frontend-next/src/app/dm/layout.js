import { Inter } from "next/font/google";
import "../globals.css";

const inter = Inter({ subsets: ["latin"] });

export const metadata = {
  title: "ChatMaps: DM",
  description: "ChatMaps: Social Media for College Students",
};

export default function RootLayout({ children }) {
  return (
    <main className={inter.className}>
      {children}
    </main>
  );
}

