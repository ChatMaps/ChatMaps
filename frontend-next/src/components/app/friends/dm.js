// Firebase Imports
import { database } from "../../../../firebase-config"
import { ref, set, get } from "firebase/database";

import ChatIcon from '@mui/icons-material/Chat';


export function openDM(user, uid) {
    var uid1 = user.uid < uid? user.uid : uid
    var uid2 = user.uid > uid? user.uid : uid
    get(ref(database, `dms/${uid1}-${uid2}`)).then((snapshot) => {
        if (snapshot.exists()) {
            window.location.href = `/dm?dm=${uid1}-${uid2}`
        } else {
            createDM(user, uid)
            window.location.href = `/dm?dm=${uid1}-${uid2}`
        }
    });
}


export function createDM(user, uid) {
    var uid1 = user.uid < uid? user.uid : uid
    var uid2 = user.uid > uid? user.uid : uid
    set(ref(database, `dms/${uid1}-${uid2}`), {
        initUID: user.uid,
        targetUID: uid,
        room: uid1 + '-' + uid2,
        UIDs: [user.uid, uid]
    })
}

/**
 * 
 * @param {JSON} friendObj - Friend Object (user) 
 * @returns DM Component
 */
export function DM({user,friendObj}) {
    return (
        <div className="border-[black] border-1 shadow-lg m-2 rounded-lg">
                <div className='grid grid-cols-4'>
                    <div className='place-content-center'>
                        <ChatIcon className='cursor-pointer' onClick={() => {openDM(user,friendObj.uid)}}/>
                    </div>
                    <div className='col-span-3 cursor-pointer'>
                        <div onClick={() => {openDM(user,friendObj.uid)}}>
                            <div className='inline-block mr-5'><img src={friendObj.pfp} className='w-[50px]'/></div>
                            <div className='inline-block relative top-[-6px]'>
                                <div className="font-bold">{friendObj.firstName} {friendObj.lastName}</div>
                                <div className="">@{friendObj.username}</div>
                            </div>
                        </div>
                    </div>
                </div>
      </div>
    )
}