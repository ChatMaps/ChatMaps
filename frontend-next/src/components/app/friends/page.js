// Dependency Imports
import { Form, useForm } from "react-hook-form";

// Firebase Imports
import { ref, set, onChildAdded, onChildRemoved } from "firebase/database";
import { database } from "../../../../firebase-config";

// Component Imports
import { Chat, SystemMessage } from "../datatypes";

// Icons
import SendIcon from '@mui/icons-material/Send';

// Notification
import { createNotification } from "../notifications/notifications";

import { useEffect, useState } from "react";

/**
 * Chat Room Component
 * @prop {JSON} roomObj - Room Object
 * @prop {JSON} user - User Object
 * @returns {Object} - Chat Room Component
 */
export function DMRoom({ roomObj, user }) {
  const [chatRoomObj, setChatRoomObj] = useState(roomObj);
  const [chats, setChats] = useState(null);
  var { register, control, reset, handleSubmit } = useForm();

  // Listeners for DMs
  useEffect(() => {
    var path = roomObj.UIDs[0] < roomObj.UIDs[1] ? roomObj.UIDs[0] + "-" + roomObj.UIDs[1] : roomObj.UIDs[1] + "-" + roomObj.UIDs[0];
    onChildAdded(ref(database, `/dms/${path}/chats`), (newDM) => {
      if (chatRoomObj) {
        var newDMRoomObj = chatRoomObj
        if (newDMRoomObj) {
          if (!newDMRoomObj.chats) {
            newDMRoomObj.chats = {}
          }
          newDMRoomObj.chats[newDM.key] = newDM.val()
          setChatRoomObj({...newDMRoomObj})
        }
      }

    });
    onChildRemoved(ref(database, `/dms/${path}/chats`), (removed) => {
      if (chatRoomObj) {
        var newDMRoomObj = chatRoomObj
        var deleted = removed.val()
        delete newDMRoomObj.chats[`${deleted.timestamp}-${deleted.user}`]
        setChatRoomObj({...newDMRoomObj})
      }
    });
  }, [])
  
  useEffect(() => {
    // Message updater
    var chatsArr = [];
    var messages = chatRoomObj.chats;
    for (var message in messages) {
      if (messages[message].isSystem) {
        chatsArr.push(
          <SystemMessage
            chatObj={messages[message]}
            key={messages[message].timestamp}
          />
        );
      } else {
        chatsArr.push(
          <Chat
            chatObj={messages[message]}
            user={user}
            path={"/dms/" + chatRoomObj.room}
            key={messages[message].timestamp}
          />
        );
      }
    }
    setChats(chatsArr.reverse())
  }, [chatRoomObj])

  /**
   * Send Message in Chatroom
   * @param {JSON} data - Message data to send (from form)
   * @returns {void}
   */
  function sendMessage(data) {
    // Other UID
    var otherUID = chatRoomObj.initUID == user.uid ? chatRoomObj.targetUID : chatRoomObj.initUID;
    // Send other user notification if not in room
    if (chatRoomObj.users && chatRoomObj.users.online) {
      if (!(otherUID in chatRoomObj.users.online)) {
        createNotification(
          "New Message",
          `${user.username} sent you a message.`,
          "dm",
          user.uid,
          otherUID
        );
      }
    }

    var messageFilterBypass = [undefined, null, "", " ", ' ', '\'']
    reset();
    if (!messageFilterBypass.includes(data.message)) {
      var payload = {
        body: data.message,
        user: user.username,
        uid: user.uid,
        isSystem: false,
        timestamp: new Date().getTime(),
      };
      set(
        ref(
          database,
          `/dms/${chatRoomObj.room}/chats/${new Date().getTime()}-${user.username}`
        ),
        payload
      );
    }
  }

  if (!chats) return <div>No Chats</div>;
  return (
    <div className="m-1 h-[100%] rounded-lg">
      <div className="h-[90%] m-4 overflow-y-auto flex flex-col-reverse">
        {chats}
      </div>
      <div className="m-2 h-[10%] w-[100%] bg-white rounded-lg">
        <Form
          onSubmit={handleSubmit(sendMessage)}
          control={control}
        >
          <div className="width-[100%] grid grid-cols-6 pr-5 pt-1">
            <input type="text" {...register("message")} placeholder="Enter message..." className="col-span-5 border-[0px]" />
            <button className="p-2 cursor-pointer bg-cyan-500 text-white font-bold rounded-full mr-5 w-[100%]"><SendIcon/></button>
          </div>
        </Form>
      </div>
    </div>
  );
}
