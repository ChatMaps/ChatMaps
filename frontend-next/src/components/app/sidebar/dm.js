import { Member } from "../datatypes"

import { database } from "../../../../firebase-config"
import {ref, get, set} from "firebase/database"

import { useState, useEffect } from "react"

export function Sidebar({user, chatRoomObj}) {
    const [profileData, setProfileData] = useState(null)
    // Active users list
    if (
      chatRoomObj.hasOwnProperty("users") &&
      chatRoomObj.users.hasOwnProperty("online")
    ) {
      var activeUsers = [];
      var activeUsersJSON = chatRoomObj.users.online;
      for (var activeUser in activeUsersJSON)
        activeUsers.push(<Member memberObj={activeUsersJSON[activeUser]} />);
      var chatroomOnline = activeUsers
    }

    useEffect(() => {
        if (user) {
            // Profile Information
            if (user.uid == chatRoomObj.UIDs[0]) {
                var profileUID = chatRoomObj.UIDs[1]
            } else {
                var profileUID = chatRoomObj.UIDs[0]
            }

            get(ref(database, `users/${profileUID}`)).then((snapshot) => {
                var profileData = snapshot.val()
                setProfileData(profileData)
            })
        }
    }, [user])
  
    return (
      <div>
        {profileData && (
        <div className="overflow-hidden h-dvh">
            <div className="m-2 h-[98%] grid grid-cols-1">
                <div className="flex place-content-center">
                    <div className="bg-white rounded-lg m-2 shadow-2xl pt-10">
                        <img src={profileData.pfp} className="w-[80%] relative mx-auto"/>
                        <div className="font-bold text-[24px]">{profileData.firstName} {profileData.lastName}</div>
                        @{profileData.username}
                    </div>
                </div>
                <div className="bg-white rounded-lg m-2 shadow-2xl">
                <div>In The Chat</div>
                {chatroomOnline}
                </div>
            </div>
          </div>
        )}
      </div>
    );
  }
  