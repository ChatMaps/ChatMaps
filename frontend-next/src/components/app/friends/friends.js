import Link from 'next/link'

// Icons
import ChatIcon from '@mui/icons-material/Chat';

// Firebase Imports
import { database } from "../../../../firebase-config"
import { ref, set } from "firebase/database";

import { openDM } from './dm';

// Icons
import CheckIcon from '@mui/icons-material/Check';
import CloseIcon from '@mui/icons-material/Close';
import CircleIcon from '@mui/icons-material/Circle';

/**
 * Send a friend request to a user
 * @param {JSON} user - User Object 
 * @param {JSON} uid - User ID of the user to send a friend request to
 */
export function addFriend(user, uid) {
    // Add to user's friend requests
    set(ref(database, `users/${uid}/friends/requests/${user.uid}`), {
        uid: user.uid
    })
}

/**
 * 
 * @param {JSON} friendObj - Friend Object (user) 
 * @returns Friend Component
 */
export function Friend({user,friendObj}) {
    return (
        <div className="border-[black] border-1 shadow-lg m-2 rounded-lg">
                <div className='grid grid-cols-4'>
                    <div className='place-content-center'>
                        <ChatIcon className='cursor-pointer' onClick={() => {openDM(user,friendObj.uid)}}/>
                    </div>
                    <div className='col-span-3 cursor-pointer'>
                        <Link href={`/user?uid=${friendObj.uid}`}>
                            <div className='inline-block mr-8'><img src={friendObj.pfp} className= 'w-[38px] h-[50px]'/></div>
                            <div className='inline-block relative top-[-6px]'>
                                <div className="font-bold">{friendObj.lastOnline == true && <CircleIcon className="text-lime-600 mr-1 relative top-[-2px]" fontSize="20px"/>}{friendObj.firstName} {friendObj.lastName}</div>
                                <div className="">@{friendObj.username}</div>
                            </div>
                        </Link>
                    </div>
                </div>
      </div>
    )
}

/**
 * 
 * @prop {JSON} user - User Object 
 * @prop {JSON} requestingUser - User Object of the user requesting to be friends
 * @returns {Object} Friend Request Component
 */
export function FriendRequest({user, requestingUser}) {
    /**
     * Accepts Friend Request
     * @param {JSON} user - User Object
     * @param {JSON} uid - User ID of the user who sent the friend request
     */
    function acceptRequest(user, uid) {
        // Add to user's friends
        set(ref(database, `users/${user.uid}/friends/friends/${uid}`), {
            uid: uid
        })
        set(ref(database, `users/${uid}/friends/friends/${user.uid}`), {
            uid: user.uid
        })
        removeRequest(user, uid)
    }
    /**
     * Removes Friend Request
     * @param {JSON} user - User Object
     * @param {JSON} uid - User ID of the user who sent the friend request
     */
    function removeRequest(user, uid) {
        // Remove from user's friend requests
        set(ref(database, `users/${user.uid}/friends/requests/${uid}`), null)
    }
    return (
        <div className="border-[black] border-1 shadow-lg m-2 rounded-lg">
            <div className='grid grid-cols-4'>
                <div className='place-content-center'>
                    <CheckIcon className='cursor-pointer ml-5' onClick={() => {acceptRequest(user, requestingUser.uid)}}/>
                    <CloseIcon className='cursor-pointer ml-5' onClick={() => {removeRequest(user, requestingUser.uid)}}/>
                </div>
                <div className='col-span-3 cursor-pointer'>
                    <Link href={`/user?uid=${requestingUser.uid}`}>
                        <div className='inline-block mr-5'><img src={requestingUser.pfp} className='w-[50px]'/></div>
                        <div className='inline-block relative top-[-6px]'>
                            <div className="font-bold">{requestingUser.firstName} {requestingUser.lastName}</div>
                            <div className="">@{requestingUser.username}</div>
                        </div>
                    </Link>
                </div>
            </div>
        </div>
    )
}