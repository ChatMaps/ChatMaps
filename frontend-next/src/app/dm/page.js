"use client";
// System Imports
import Drawer from '@mui/material/Drawer';
import { useState, useEffect } from "react";

// Firebase Imports
import { auth, database } from "../../../firebase-config";
import { ref, onValue, set, onDisconnect  } from "firebase/database";
import { useAuthState } from "react-firebase-hooks/auth"

// Component Imports
import { Header } from "../../components/app/header";
import { DMRoom } from "../../components/app/friends/page";
import { Sidebar } from "../../components/app/sidebar/dm";
import {useWindowSize} from "../../components/app/datatypes";


/**
 * DM Page
 * @returns {Object} Chat Page
 */
function Chat() {
  // State variables for chat page
  const [user, setUser] = useState(null); // user data
  const [chatRoomObj, setChatRoomObj] = useState(null); // Current chatroom object
  const [doneLoading, setDoneLoading] = useState(false) // is the page done loading or not
  const [authUser] = useAuthState(auth) // auth user object (used to obtain other user object)
  const [drawerOpen, setDrawerOpen] = useState(true); // drawer open state

  var windowSize = useWindowSize()
  useEffect(() => {
    if (windowSize.width < 767) {
      setDrawerOpen(false)
    } else {
      setDrawerOpen(true)
    }
  }, [windowSize])
  
  // Authentication Verification / Redirection if Profile Data not Filled out
  useEffect(() => {
    if (authUser) {
        onValue(ref(database, `users/${authUser.uid}`), (userData) => {
            userData = userData.val();
            if (userData) {
                setUser(userData);
            } else {
                window.location.href = "/onboarding";
            }
        });
    }
  }, [authUser])

  // Users URL params to load proper chatroom, then logs the user into that room
  useEffect(() => {
    if (user) {
        const searchParams = new URLSearchParams(document.location.search);
        var path = searchParams.get("dm")

        /*// Send entered message
        var payload = {
            body: "entered",
            user: user.username,
            isSystem: true,
            timestamp: new Date().getTime(),
            uid: user.uid,
        };
        set(
            ref(
            database,
            `/rooms/${path}/chats/${new Date().getTime()}-${user.username}`
            ),
            payload
        );*/

        // Add user to online for room
        set(ref(database, `/dms/${path}/users/online/${user.uid}`), user)

        // Removes user from room on disconnect (reload, window close, internet lost)
        onDisconnect(ref(database, `/dms/${path}/users/online/${user.uid}`)).remove()

        // Sends leaving message on disconnect (Timestamp function used due to new onDisconnect stuff)
        /*someRef = ref(database, `/rooms/${path}/chats/${new Date().getTime()}-${user.username}`)
        onDisconnect(someRef).set({
          body: "left",
          user: user.username,
          isSystem: true,
          timestamp: serverTimestamp(),
          uid: user.uid,
        })*/

        onValue(ref(database, `/dms/${path}`), (roomData) => {
            roomData = roomData.val();
            setChatRoomObj(roomData)
            if (!doneLoading) {
                setDoneLoading(true)
            }
        })
    }
  }, [user]);

  return (
    <div>
      {(authUser && doneLoading) && (
        <div className="overflow-hidden h-dvh">
          {/* Left Side of Page */}
          <div className="overflow-hidden h-dvh md:mr-[400px]">
            {/* Header */}
            <Header
              mainTab={"dm"}
              chatRoomObj={chatRoomObj}
              user={user}
              sidebarControl={() => {setDrawerOpen(!drawerOpen)}}
            />
            {/* Main Page Section */}
            <div className="mr-2 h-[calc(100%-110px)]">
                <DMRoom roomObj={chatRoomObj} user={user} />
            </div>
          </div>
          {/* Sidebar (Right Side of Page) */}
            <Drawer open={drawerOpen} anchor={"right"} variant={windowSize.width > 767? "persistent": "temporary"} onClose={() => {setDrawerOpen(false)}} sx={{
          width: windowSize.width > 767? 400: "80%",
          marginTop: 10,
          flexShrink: 0,
          '& .MuiDrawer-paper': {
            width: windowSize.width > 767? 400: "80%",
            borderLeft: 0,
          },
          }}>
            <div className="shadow-2xl">
              <Sidebar chatRoomObj={chatRoomObj} user={user}/>
            </div>
          </Drawer>
        </div>
      )}
    </div>
  );
}

export default Chat;