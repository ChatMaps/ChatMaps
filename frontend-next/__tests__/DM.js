import { DM } from "../src/components/app/friends/dm";
import renderer from 'react-test-renderer';

var message = "Hello, World! This is a test message. https://www.google.com"

var exampleUser = {
    uid: "123456",
    username: "TestUser",
    firstName: "Test",
    lastName: "User",
    lastOnline: true,
    pfp: "https://th.bing.com/th/id/OIP.K5VoTfw97JiEc1OBODAjmQHaHO?rs=1&pid=ImgDetMain"
}

var friendObj = {
    uid: "123456",
    username: "TestUser",
    firstName: "Test",
    lastName: "Friend",
    lastOnline: true,
    pfp: "https://th.bing.com/th/id/OIP.K5VoTfw97JiEc1OBODAjmQHaHO?rs=1&pid=ImgDetMain"

}

it('RMF Renders Correctly', () => {
    const tree = renderer
        .create(<DM user={exampleUser} friendObj={friendObj} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});