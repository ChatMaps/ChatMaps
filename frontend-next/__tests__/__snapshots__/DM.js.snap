// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RMF Renders Correctly 1`] = `
<div
  className="border-[black] border-1 shadow-lg m-2 rounded-lg"
>
  <div
    className="grid grid-cols-4"
  >
    <div
      className="place-content-center"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium cursor-pointer css-i4bv87-MuiSvgIcon-root"
        data-testid="ChatIcon"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2M6 9h12v2H6zm8 5H6v-2h8zm4-6H6V6h12z"
        />
      </svg>
    </div>
    <div
      className="col-span-3 cursor-pointer"
    >
      <div
        onClick={[Function]}
      >
        <div
          className="grid grid-cols-2 justify-items-center"
        >
          <div
            className="mr-8"
          >
            <img
              className="w-[50px] h-[50px]"
              src="https://th.bing.com/th/id/OIP.K5VoTfw97JiEc1OBODAjmQHaHO?rs=1&pid=ImgDetMain"
            />
          </div>
          <div
            className=""
          >
            <div
              className="font-bold"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSize20px text-lime-600 mr-1 relative top-[-2px] css-821wyw-MuiSvgIcon-root"
                data-testid="CircleIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2"
                />
              </svg>
              Test
               
              Friend
            </div>
            <div
              className=""
            >
              @
              TestUser
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
